File: ios/zing_sdk_initializer/Sources/zing_sdk_initializer/CriticalErrorHandler.swift
import Flutter
import ZingCoachSDK

final class CriticalErrorHandler: ZingCoachSDK.CriticalErrorHandler {
    private let channel: FlutterMethodChannel

    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }

    func sdkDidFail(with error: Error) {
        let code = (error as? FlutterCodableError)?.flutterCode ?? "unknown"
        DispatchQueue.main.async { [channel] in
            channel.invokeMethod(
                "onCriticalError",
                arguments: [
                    "code": code,
                    "message": String(describing: error)
                ]
            )
        }
    }
}

File: android/src/main/kotlin/com/example/zing_sdk_initializer/CriticalErrorHandler.kt
package com.example.zing_sdk_initializer

import coach.zing.fitness.coach.auth.AuthError
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/** Forwards native SDK critical errors to Dart over [channel]. */
class CriticalErrorHandler(
    private val channel: MethodChannel,
    private val scope: CoroutineScope,
) : coach.zing.fitness.coach.CriticalErrorHandler {

    override fun onSessionExpired(error: AuthError) {
        scope.launch {
            channel.invokeMethod(
                "onCriticalError",
                mapOf(
                    "code" to "auth_error",
                    "message" to error.toString(),
                ),
            )
        }
    }
}
